### PR TITLE
update to expose contexts

### PR DESCRIPTION
**Pull request by gregsdennis**
Really just want to show the Logic lib.  The rest isn't quite ready.

This showcases that we can expose the contexts, which gives the user access to the JTIs for all of the types contained in or used by the lib.  We still use the options manager to handle deserializing user-registered rules, but it's kept internal.  The hard part is remembering on which context your JTI exists.

Also, because the context is exposed, there's no need to have the `IJsonTypeInfoResolver` static property.

Ultimately, I see this working out for all the libs.  We can also create a serializer context in Json.More that exposes common types, like JsonNode.  (Really, I think .Net should expose a context for its common types... \*cough\* @eiriktsarpalis)

We might also be able to drop the `IJsonConverterReadWrite`.

### CODE REVIEW

File: JsonSchema.OpenApi/DiscriminatorKeyword.cs [hunk 1]
```diff
@@ -55,7 +55,7 @@ public DiscriminatorKeyword(string propertyName, IReadOnlyDictionary<string, str
 		Mapping = mapping;
 		Extensions = extensions;
 
-		_json = JsonSerializer.SerializeToNode(this, OpenApiSerializerContext.OptionsManager.SerializerOptions);
+		_json = JsonSerializer.SerializeToNode(this, OpenApiSerializerContext.Default.JsonNode);
```
Comment (jevansaks) [inline JsonSchema.OpenApi/DiscriminatorKeyword.cs:58]: Why is this JsonNode and not DiscriminatorKeyword? This seems like a mismatch

Comment (gregsdennis) [reply JsonSchema.OpenApi/DiscriminatorKeyword.cs:58]: oh... ignore that.  Just look at the logic stuff.  That's the only thing guaranteed to work so far.

File: JsonLogic/Rule.cs [hunk 2]
```diff
@@ -119,9 +111,11 @@ public override Rule Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSer
 				if (ruleType == null)
 					throw new JsonException($"Cannot identify rule for {op}");
 
+				var typeInfo = LogicSerializerContext.OptionsManager.SerializerOptions.GetTypeInfo(ruleType);
```
Comment (jevansaks) [inline JsonLogic/Rule.cs:114]: Let's just go straight to the RuleRegistry's lookup table and then we don't need to go through the ExternalTypeInfoResolvers thing. We might even be able to get rid of the OptionsManager.

Comment (gregsdennis) [reply JsonLogic/Rule.cs:114]: I'll give that a go.

**Comment by jevansaks** [comment]
> 		where T : Rule

Now that we are targeting latest System.Text.Json, we can change this parameter on `AddRule<T>` and `RegisterKeyword<T>`  to be an IJsonTypeInfoResolver. I had only made it a JsonSerializerContext because STJ6 didn't have the interface.

---
Refers to: JsonLogic/RuleRegistry.cs:118 in 075c705. [](commit_id = 075c7050a4cc907bf886a955a13a8fe3e2bdbcee, deletion_comment = False)

File: JsonSchema/SchemaKeywordRegistry.cs [hunk 3]
```diff
@@ -102,6 +102,7 @@ static SchemaKeywordRegistry()
 		};
 
 		_keywords = new ConcurrentDictionary<string, Type>(keywordData.ToDictionary(x => x.Item2, x => x.Item1));
+		_keywordTypeInfoResolvers = new ConcurrentDictionary<Type, JsonSerializerContext>(keywordData.ToDictionary(x => x.Item1, _ => (JsonSerializerContext)JsonSchemaSerializerContext.Default));
```
Comment (jevansaks) [inline JsonSchema/SchemaKeywordRegistry.cs:106]: RuleRegistry did this differently and just returns null for known types. It seems more efficient to not populate this dictionary and just return the default context when the keyword type isn't found.

File: JsonSchema/SchemaKeywordRegistry.cs [hunk 4]
```diff
@@ -173,6 +172,13 @@ public static void Unregister<T>()
 			: null;
 	}
 
+	internal static JsonTypeInfo GetTypeInfo(Type ruleType)
+	{
+		return _keywordTypeInfoResolvers.TryGetValue(ruleType, out var context)
+			? context.GetTypeInfo(ruleType)!
+			: JsonSchemaSerializerContext.Default.GetTypeInfo(typeof(UnrecognizedKeyword))!;
+	}
```
Comment (jevansaks) [inline JsonSchema/SchemaKeywordRegistry.cs:180]: Why is this ignoring ruleType here and passing UnrecognizedKeyword explicitly?

File: JsonSchema.Data/Vocabularies.cs [hunk 5]
```diff
@@ -31,27 +30,29 @@ public static void Register(VocabularyRegistry? vocabRegistry = null, SchemaRegi
...
+	internal static JsonSerializerOptions CombinedOptions { get; }
 
-	static DataExtSerializerContext() {
-		OptionsManager = new(
-			Default,
-			JsonSchema.TypeInfoResolver
-		);
+	static JsonSchemaDataSerializerContext()
+	{
+		CombinedOptions = new JsonSerializerOptions
```
Comment (jevansaks) [inline JsonSchema.Data/Vocabularies.cs:53]: Why do we need to combine options anymore? This new strategy should avoid the need for it.

File: JsonSchema.Data.Tests/Suite/Suite.cs [hunk 6]
```diff
@@ -43,10 +55,7 @@ public static IEnumerable<TestCaseData> TestCases()
 		{
 			var shortFileName = System.IO.Path.GetFileNameWithoutExtension(fileName);
 			var contents = File.ReadAllText(fileName);
-			var collections = JsonSerializer.Deserialize<List<TestCollection>>(contents, new JsonSerializerOptions(DataTestsSerializerContext.OptionsManager.SerializerOptions)
-			{
-				PropertyNameCaseInsensitive = true
-			});
+			var collections = JsonSerializer.Deserialize<List<TestCollection>>(contents, _serializerOptions);
```
Comment (jevansaks) [inline JsonSchema.Data.Tests/Suite/Suite.cs:58]: Similarly, this call should be able to just pass DataTestsSerializerContext.Default.ListTestCollection, right?

File: Json.More/AotCompatibleJsonConverter.cs [hunk 7]
```diff
@@ -21,7 +22,8 @@ public interface IJsonConverterReadWrite
 		/// <param name="writer">The writer to write to.</param>
 		/// <param name="value">The value to convert to JSON.</param>
 		/// <param name="options">An object that specifies serialization options to use.</param>
-		void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options);
+		/// <param name="typeInfo">The <see cref="JsonTypeInfo"/> for the value being serialized.</param>
+		void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options, JsonTypeInfo? typeInfo = null);
```
[the remark this answers is not in the document]
Comment (gregsdennis) [reply Json.More/AotCompatibleJsonConverter.cs:26]: Read is working.  I didn't need to update it.  I can look at doing that, though.

File: JsonSchema/SchemaKeywordRegistry.cs (hunk 4 above)
Comment (gregsdennis) [reply JsonSchema/SchemaKeywordRegistry.cs:180]: If the keyword isn't found, it's treated as an unrecognized keyword.  This captures the data for the keyword.  It doesn't play into validation and just generates and annotation.

File: JsonSchema.Data/Vocabularies.cs (hunk 5 above)
Comment (gregsdennis) [reply JsonSchema.Data/Vocabularies.cs:53]: This was the workaround for the bug.  I need it in the `UriIdentifier` class where I'm serializing a `JsonSchema`.

File: JsonSchema.Data.Tests/Suite/Suite.cs (hunk 6 above)
Comment (gregsdennis) [reply JsonSchema.Data.Tests/Suite/Suite.cs:58]: I wanted to be able to set the options here.

File: JsonSchema.Data/Vocabularies.cs (hunk 5 above)
Comment (jevansaks) [reply JsonSchema.Data/Vocabularies.cs:53]: Which bug? Can you include a comment / pointer to the issue if it's a bug in STJ?

**Comment by gregsdennis** [comment]
closing in favor of #648 which includes these changes.

**Pull request closed without merging**